File: cp35/python/graphs/dfs.py

def dfs_util(adj, visited, current):
  if visited[current]:
    return
  print(current, "->", end=" ")
  visited[current] = True
  for neighbor in adj[current]:
    if not visited[neighbor]:
      print(neighbor, end=" ")
      dfs_util(adj, visited, neighbor)
  print()

def dfs(adj):
  n = len(adj)
  visited = [0] * 100
  for i in range(n):
    dfs_util(adj, visited, i)


if __name__ == "__main__":
  
  dfs([[1, 2], [0, 3], [0, 4], [3, 2], [4, 3]])

File: cp35/cpp/graphs/dfs.cpp
#include<bits/stdc++.h>
using namespace std;

void dfs_util(vector<vector<int>> &adj, vector<bool> &visited, int current){
  if(visited[current]){
    return;
  }
  cout<<current<<" -> ";
  visited[current] = 1;
  for(auto neighbor: adj[current]){
    if(!visited[neighbor]){
      cout<< neighbor << " " << endl;
      dfs_util(adj, visited, neighbor);
    }
  }
  cout << endl;
}

void dfs(vector<vector<int>> adj){
  int n = adj.size();
  vector<bool> visited(n);
  for(int i = 0; i < n; i++){
    dfs_util(adj, visited, i);
  }
}

int main(){
      dfs({{1, 4},
         {0, 2, 3},
         {1, 3},
         {1, 2},
         {0}});
    
      dfs({{1, 4},
         {0},
         {3},
         {2},
         {0}});
}